### ISSUE TITLE
Make `load_font_data` return `Result`?

### ISSUE THREAD

**Issue by si14** [RazrFalcon/fontdb#76]
Recently, I spent a few hours trying to figure out why `resvg` renders completely fine on my laptop, but doesn't draw anything on the server with zero failures. The root cause was a forgotten `git lfs` checkout, but my mistake wasn't obvious because of the way `fontdb` handles font loading.

Here is how I used it:

`fontdb.load_font_data(include_bytes!(...))`

It looks perfectly safe, the file is present, but unfortunately `fontdb` has this bit:

```rust
                match parse_face_info(source.clone(), data, index) {
                    Ok(mut info) => {
                        let id = self.faces.insert_with_key(|k| {
                            info.id = ID(k);
                            info
                        });
                        ids.push(ID(id));
                    }
                    Err(e) => log::warn!(
                        "Failed to load a font face {} from source cause {}.",
                        index,
                        e
                    ),
                }
```

In other words, font loading is silently swallowing errors. `log::warn!` is helpful, but is not outputting anything if logging isn't configured.

Would you be open to changing the API and making all font loading functions return `Result`? It won't be, strictly speaking, a breaking change, but it will cause "unused `Result`" warnings in clients.

I'm happy to take a stab on a PR if you agree it's a good idea.

**Comment by RazrFalcon** [comment]
>log::warn! is helpful, but is not outputting anything if logging isn't configured.

Well, then it's a caller problem.

>Would you be open to changing the API and making all font loading functions return Result?
>It won't be a breaking change

It will be. Not to mention that the current logic skips malformed font faces and doesn't bail immediately. So returning an error, or rather a list of errors for each face would be rather problematic and would require a lot of changes.

Either way, this crate is unmaintained. So there will be no new versions.

**Comment by si14** [comment]
> Well, then it's a caller problem.

Sure, but I believe it's not very common for Rust crates to rely on logging to signal critical errors (arguably, not being able to load a request font is a critical error).

> Not to mention that the current logic skips malformed font faces and doesn't bail immediately. So returning an error, or rather a list of errors for each face would be rather problematic and would require a lot of changes.

I see. Would you be opposed to a separate method (something like `_strict`)? This should be completely reverse compatible.

> Either way, this crate is unmaintained. So there will be no new versions.

Oh no! Is there a succession plan for `resvg`/`usvg`? Does that include ready-to-publish PRs? Also, would you like me to submit a README.md PR with a notice?

**Comment by RazrFalcon** [comment]
>Sure, but I believe it's not very common for Rust crates to rely on logging to signal critical errors

The current API is designed for loading system fonts, in which case an error is highly unlikely. 

>This should be completely reverse compatible.

The problem is not in breaking the API, but in finding a good solution for this problem. And because this crate is unmaintained it would not happen any time soon.

>Is there a succession plan for resvg/usvg?

`resvg` is maintained by [linebender](https://github.com/linebender) now.

**Comment by si14** [comment]
Thank you, I brought it up in the resvg repo https://github.com/linebender/resvg/issues/862

I hope you are well (as much as it's possible in the circumstances), and again thanks a lot for your work. I hope a donation to Повернись живим works as a tiny token of my gratitude.
